sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala

```diff
@@ -0,0 +1,237 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.streaming
+
+import java.time.Duration
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.Encoder
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchema.{KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA_WITH_TTL}
+import org.apache.spark.sql.execution.streaming.state.{NoPrefixKeyStateEncoderSpec, StateStore, StateStoreErrors}
+import org.apache.spark.sql.streaming.{TTLMode, ValueState}
+
+/**
+ * Class that provides a concrete implementation for a single value state associated with state
+ * variables (with ttl expiration support) used in the streaming transformWithState operator.
+ *
+ * @param store - reference to the StateStore instance to be used for storing state
+ * @param stateName - name of logical state partition
+ * @param keyExprEnc - Spark SQL encoder for key
+ * @param valEncoder - Spark SQL encoder for value
+ * @param ttlMode    - TTL Mode for values  stored in this state
+ * @param batchTimestampMs - processing timestamp of the current batch.
+ * @param eventTimeWatermarkMs - event time watermark for streaming query
+ *                               (same as watermark for state eviction)
+ * @tparam S - data type of object that will be stored
+ */
+class ValueStateImplWithTTL[S](
+    store: StateStore,
+    stateName: String,
+    keyExprEnc: ExpressionEncoder[Any],
+    valEncoder: Encoder[S],
+    ttlMode: TTLMode,
+    batchTimestampMs: Option[Long],
+    eventTimeWatermarkMs: Option[Long])
+  extends ValueState[S] with Logging with StateVariableWithTTLSupport {
+
+  private val keySerializer = keyExprEnc.createSerializer()
+  private val stateTypesEncoder = StateTypesEncoder(keySerializer, valEncoder,
+    stateName, hasTtl = true)
+  private[sql] var ttlState: SingleKeyTTLStateImpl = _
+
+  initialize()
+
+  private def initialize(): Unit = {
+    assert(ttlMode != TTLMode.NoTTL())
+
+    store.createColFamilyIfAbsent(stateName, KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA_WITH_TTL,
+      NoPrefixKeyStateEncoderSpec(KEY_ROW_SCHEMA))
+
+    ttlState = new SingleKeyTTLStateImpl(ttlMode, stateName, store,
```

Comment:
Could we pass `this` here ?

Reply:
Modified to use one trait `TTLState` and extend it from state variable.